Convert an ucs4 string to the current locale instead of to converting it
to "latin1".  This should make it possible to display it on an UTF-8 locale.

--- madplay-0.15.2b.orig/player.c
+++ madplay-0.15.2b/player.c
@@ -28,6 +28,8 @@
 # include <stdio.h>
 # include <stdarg.h>
 # include <stdlib.h>
+# include <wchar.h>
+
 
 # ifdef HAVE_SYS_TYPES_H
 #  include <sys/types.h>
@@ -598,6 +600,77 @@
   return 0;
 }
 
+/* Note: id3_ucs4_length and id3_ucs4_size are duplicated from
+ * libid3tag because they are not exported. */
+
+/*
+ * NAME:        ucs4->length()
+ * DESCRIPTION: return the number of ucs4 chars represented by a ucs4 string
+ */
+id3_length_t id3_ucs4_length(id3_ucs4_t const *ucs4)
+{
+	id3_ucs4_t const *ptr = ucs4;
+
+	while (*ptr)
+		++ptr;
+
+	return ptr - ucs4;
+}
+
+/*
+ * NAME:        ucs4->size()
+ * DESCRIPTION: return the encoding size of a ucs4 string
+ */
+id3_length_t id3_ucs4_size(id3_ucs4_t const *ucs4)
+{
+	  return id3_ucs4_length(ucs4) + 1;
+}
+
+/* NAME:	ucs4_to_wchar_strdup()
+ * DESCRIPTION: convert an ucs4 string to a wchar_t string
+ */
+static
+wchar_t	*ucs4_to_wchar_strdup(const id3_ucs4_t *ucs4)
+{
+	id3_length_t	size = id3_ucs4_size(ucs4);
+	wchar_t		*s = malloc(size * sizeof(wchar_t));
+	id3_length_t	i;
+
+	for (i = 0; i < size; i++)
+	{
+		s[i] = ucs4[i];
+	}
+	return s;
+}
+
+/*
+ * NAME:	ucs4_to_locale_strdup()
+ * DESCRIPTION:	convert ucs4 string to locale encoding
+ */
+static
+char *ucs4_to_locale_strdup(const id3_ucs4_t *ucs4)
+#ifndef HAVE_WCSLEN
+{
+  return id3_ucs4_latin1duplicate(ucs4);
+}
+#else
+{
+  /*
+   * UTF8 is at most 5 bytes per character and should be the worst
+   * case encoding, so we use this here.
+   */
+  wchar_t	*ws = ucs4_to_wchar_strdup(ucs4);	
+  int len = wcslen(ws)*5;
+  char *s = malloc(len);
+
+  snprintf(s, len, "%ls", ws);
+  s[len - 1] = 0;
+  free(ws);
+
+  return s;
+}
+#endif
+
 /*
  * NAME:	show_id3()
  * DESCRIPTION:	display ID3 tag information
@@ -652,7 +725,7 @@
       if (strcmp(info[i].id, ID3_FRAME_GENRE) == 0)
 	ucs4 = id3_genre_name(ucs4);
 
-      latin1 = id3_ucs4_latin1duplicate(ucs4);
+      latin1 = ucs4_to_locale_strdup(ucs4);
       if (latin1 == 0)
 	goto fail;
 
@@ -685,7 +758,7 @@
     ucs4 = id3_field_getfullstring(id3_frame_field(frame, 3));
     assert(ucs4);
 
-    latin1 = id3_ucs4_latin1duplicate(ucs4);
+    latin1 = ucs4_to_locale_strdup(ucs4);
     if (latin1 == 0)
       goto fail;
 
@@ -1211,7 +1284,7 @@
     if (nstrings > 0) {
       id3_latin1_t *latin1;
 
-      latin1 = id3_ucs4_latin1duplicate(id3_field_getstrings(field, 0));
+      latin1 = ucs4_to_locale_strdup(id3_field_getstrings(field, 0));
       if (latin1) {
 	signed long ms;
 
